src/solvers/flattening/pointer_logic.cpp

```diff
@@ -23,13 +23,11 @@ Author: Daniel Kroening, kroening@kroening.com
 
 bool pointer_logict::is_dynamic_object(const exprt &expr) const
 {
-  if(expr.type().get_bool(ID_C_dynamic))
-    return true;
-
-  return expr.id() == ID_symbol &&
+  const bool is_dynamic = expr.type().get_bool(ID_C_dynamic);
+  const auto symbol_expr = expr_try_dynamic_cast<symbol_exprt>(expr);
```

Comment:
Apart from the bug that @romainbrenguier spotted, this approach has the downside that we do extra work that may never be necessary in case `is_dynamic` is true. Could we just do
```
return expr.type().get_bool(ID_C_dynamic) || (... return statement from the first commit ...);
```
?

I don't think this is super performance critical, but I don't think this is a massive improvement either.